mava/systems/q_learning/anakin/rec_qmix.py

```diff
@@ -0,0 +1,697 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import time
+from typing import Any, Dict, Tuple
+
+import chex
+import flashbax as fbx
+import hydra
+import jax
+import jax.lax as lax
+import jax.numpy as jnp
+import optax
+from colorama import Fore, Style
+from flashbax.buffers.flat_buffer import TrajectoryBuffer
+from flax.core.scope import FrozenVariableDict
+from flax.linen import FrozenDict
+from jax import Array, tree
+from jumanji.types import TimeStep
+from omegaconf import DictConfig, OmegaConf
+from rich.pretty import pprint
+
+from mava.evaluator import ActorState, get_eval_fn, get_num_eval_envs
+from mava.networks import RecQNetwork, ScannedRNN
+from mava.networks.base import QMixNetwork
+from mava.systems.q_learning.types import (
+    ActionSelectionState,
+    ActionState,
+    LearnerState,
+    Metrics,
+    QMIXParams,
+    TrainState,
+    Transition,
+)
+from mava.types import MarlEnv, Observation
+from mava.utils import make_env as environments
+from mava.utils.checkpointing import Checkpointer
+from mava.utils.jax_utils import (
+    switch_leading_axes,
+    unreplicate_batch_dim,
+    unreplicate_n_dims,
+)
+from mava.utils.logger import LogEvent, MavaLogger
+from mava.utils.total_timestep_checker import check_total_timesteps
+from mava.wrappers import episode_metrics
+
+
+def init(
+    cfg: DictConfig,
+) -> Tuple[
+    Tuple[MarlEnv, MarlEnv],
+    RecQNetwork,
+    QMixNetwork,
+    optax.GradientTransformation,
+    TrajectoryBuffer,
+    LearnerState,
+    MavaLogger,
+    chex.PRNGKey,
+]:
+    """Initialize system by creating the envs, networks etc."""
```

Comment:
Not too fussed if you remove this docstring, but if you remove it from here then please do the same for IQL